### PR TITLE
[MRG+1] FIX: Only render single patch for scatter

### PR DESCRIPTION
**Pull request by tacaswell**
Alternative to #6883

### CODE REVIEW

File: lib/matplotlib/legend_handler.py [hunk 1]
```diff
@@ -316,7 +316,7 @@ def get_sizes(self, legend, orig_handle,
             numpoints = self.get_numpoints(legend)
             if numpoints < 4:
                 sizes = [.5 * (size_max + size_min), size_max,
```
Comment (NelleV) [inline lib/matplotlib/legend_handler.py]: That could easily be unit tested.

File: lib/matplotlib/backend_bases.py [hunk 2]
```diff
@@ -400,7 +400,7 @@ def _iter_collection_uses_per_path(self, paths, all_transforms,
         if Npaths == 0 or (len(facecolors) == 0 and len(edgecolors) == 0):
             return 0
         Npath_ids = max(Npaths, len(all_transforms))
-        N = max(Npath_ids, len(offsets))
+        N = max(Npaths, len(offsets))
```
Comment (NelleV) [inline lib/matplotlib/backend_bases.py]: If I am not mistaken, this function is only used to compute whether or not to do an optimization. Hence, over estimating the number of paths drawn is not that important and I am assuming that this as no impact on the bug you are fixing?

**Comment by efiring** [comment]
The difference in the test_legend png baseline plots is imperceptible by eye.  Perhaps it would be worthwhile to modify that test to include the alpha and size variation from the scatter example that showed the problem in the first place.

File: lib/matplotlib/legend_handler.py (hunk 1 above)
Comment (tacaswell) [inline lib/matplotlib/legend_handler.py]: It is possible that _this_ is the only change that really matters here.

File: lib/matplotlib/backend_bases.py (hunk 2 above)
Comment (anntzer) [inline lib/matplotlib/backend_bases.py]: (Also, if we put a fix there, I'd also just replace the next line by `return int(np.ceil(N / Npath_ids))` which is much clearer in intent.)

File: lib/matplotlib/legend_handler.py (hunk 1 above)
Comment (NelleV) [inline lib/matplotlib/legend_handler.py]: all the more reason to unit test it, no?
It is a very easy unit test to check that the return lengths is equal to numpoints.

Also, my two cents is that I would much prefer the returned size to be sorted (only matters for numpoints == 3)

**Comment by NelleV** [comment]
After reviewing with @anntzer today, I finally understand what's going on.
(@anntzer I am not reproducing the bug we noticed on your computer).

**Comment by anntzer** [comment]
@NelleV I figured it out: that's because I used `range(10)` for sizes, which means that the smallest marker has size zero... putting a nonzero smallest size fixes the issue we were looking at.

File: lib/matplotlib/legend_handler.py (hunk 1 above)
Comment (anntzer) [inline lib/matplotlib/legend_handler.py]: I also vote for drawing the markers in increasing size.  (This also makes the code simpler; now only numpoints=1 needs to be special-cased.)

**Comment by NelleV** [comment]
Interesting. Maybe the code should raise a warning?

**Comment by anntzer** [comment]
@NelleV that's a bit a side issue but anyways... Currently `scatter` mostly lets negative sizes through, a warning is simply given by numpy when square-rooting them:
```
In [1]: plt.scatter(range(10), range(10), -np.arange(10))
/usr/lib/python3.5/site-packages/matplotlib/collections.py:862: RuntimeWarning: invalid value encountered in sqrt
  scale = np.sqrt(self._sizes) * dpi / 72.0 * self._factor
Out[1]: <matplotlib.collections.PathCollection at 0x7fa0f18e3dd8>
```
(resulting plot is empty).

If anything this should raise a ValueError (and whether to error on zero sizes or not is a separate issue).

File: lib/matplotlib/backend_bases.py (hunk 2 above)
Comment (anntzer) [inline lib/matplotlib/backend_bases.py]: It's just an optimization (see docstring) and it doesn't even really matter whether this is correct or not (modulo an unnecessary increase in the size of the svg).

**Comment by QuLogic** [comment]
Ping @mdboom for thoughts vs. #6883.

**Comment by NelleV** [comment]
The failure seems unrelated, but is consistently here…

**Comment by tacaswell** [comment]
.... and only on pytest.

When in doubt ~~reboot~~ rebase

**Comment by NelleV** [comment]
Thanks @tacaswell !

**Comment by tacaswell** [comment]
backported to v2.x as 568b523

**Pull request merged**